docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,341 @@
+
+.. math:: \mathrm{proc}\, 0: [0,n) \times [0,N) \times [0,K),\quad \mathrm{proc}\, 1: [n,2n) \times [0,N) \times [0,K),\quad \mathrm{proc}\, 2: [2n,N) \times [0,N) \times [0,K).
+
+.. code-block:: python
+
+    with region(WestEdge(0, 1), EastEdge(0, 1)), interval(...):
+
+This part of the example neatly indicates that it is intended to be run on both the east and west edges of horizontal plane for each `k`, with an index strip `1` wide.
+The construction and evaluation of these application-specific objects, `WestEdge(0,1)` and `EastEdge(0,1)`, are done before code generation, and can call into user-provided code.
+These objects have attributes `local_origin` and `extent` which define the iteration space in the horizontal plane `i` and `j`.
+The union of the boxes defined by each of these objects is used for iteration.
+
+When GT4Py asks for the iteration space from the region objects, it sees different `local_origin` and `extent` on each process using the local iteration space indices:
+
+.. code-block:: python
+
+    proc 0:    local_origin=(0, 0), extent=(1, N)
+    proc 1:    local_origin=(0, 0), extent=(0, 0)
+    proc 2:    local_origin=(n-1, 0), extent=(1, N)
+
+The result of the evaluation is inserted directly into code, so GT4Py and its backend never deal with partitioning.
+Translation to the local indices is done within the Box object specialization from the application.
+
+Note that if this example had only one process, the local and global iteration spaces coincide, so the `local_origin` and `extent` set simply get passed on to GT4Py unmodified.
+
+
+Compilation-Time Evaluation
++++++++++++++++++++++++++++
+
+The expression `expr` inside `region(expr)` is evaluated at compile-time, and the resulting index ranges are stored in the library module created.
+Importantly, this means that GT4Py does not need to create objects to mirror these expression at call-time, as they may call partitioners or other application-specific code.
+
+
+Region Objects
+++++++++++++++
+
+The attributes provided by the region objects, `local_origin` and `extent`, mirror the existing implementation which accepts `origin` and `domain`.
+This stays compatible with GridTools boundary conditions.
+If the `extent` is of size `0` in any dimension, the block will not execute.
+This information could be utilized by the backend to optimize not computing unnecessary stencils.
+Note that `region` takes in an arbitrary number of arguments (`region(*args)`) so the same sub-stencil can execute on any number of rectangular areas specified, as was shown in the motivating example.
+
+Application developers will create their own specializations of Box for their particular uses.
+For example, since our use case has edges and corners, we may have specific `Corner` specializations as well.
+
+When using `region()`, the stencil caller should avoid using `origin` and `domain` as this would change the iteration space and would be very error prone.
```

Comment:
should these concepts still exist though? It's not very pythonic to have two ways of doing the same thing and it seems that all the extent / origin things can be handled like this? Unless you want to apply the same stencil in different regions?

Reply:
I think the idea was for backwards compatibility and so people don't necessarily need to use this new feature. The old way could be deprecated if the gt4py developers agree this should replace the old way (would need input from other groups to see if it suits their cases).

Reply:
I don't think they should exist, but we can pose this as purely backward-compatible if we keep it for now. Let's see what the GT4Py devs think first.